pdns/opensslsigners.cc

```diff
@@ -910,6 +913,239 @@ void OpenSSLECDSADNSCryptoKeyEngine::fromPublicKeyString(const std::string& inpu
+    if (d_edkey == nullptr) {
+      throw runtime_error(getName()+" allocation of key structure failed");
+    }
+
+#ifdef HAVE_LIBCRYPTO_ED25519
+    if(d_algorithm == 15) {
+      d_len = 32;
+      d_id = NID_ED25519;
+    }
+#endif
+#ifdef HAVE_LIBCRYPTO_ED448
+    if (d_algorithm == 16) {
+      d_len = 57;
+      d_id = NID_ED448;
+    }
+#endif
+    if (d_len == 0) {
+      EVP_PKEY_free(d_edkey);
+      throw runtime_error(getName()+" unknown algorithm "+std::to_string(d_algorithm));
+    }
+  }
+
+  ~OpenSSLEDDSADNSCryptoKeyEngine()
+  {
+    EVP_PKEY_free(d_edkey);
+  }
+
+  string getName() const override { return "OpenSSL EDDSA"; }
+  int getBits() const override { return d_len; }
+
+  void create(unsigned int bits) override;
+  storvector_t convertToISCVector() const override;
+  std::string sign(const std::string& hash) const override;
+  bool verify(const std::string& msg, const std::string& signature) const override;
+  std::string getPubKeyHash() const override;
+  std::string getPublicKeyString() const override;
+  void fromISCMap(DNSKEYRecordContent& drc, std::map<std::string, std::string>& stormap) override;
+  void fromPublicKeyString(const std::string& content) override;
+  bool checkKey() const override;
+
+  static std::shared_ptr<DNSCryptoKeyEngine> maker(unsigned int algorithm)
+  {
+    return std::make_shared<OpenSSLEDDSADNSCryptoKeyEngine>(algorithm);
+  }
+
+private:
+  size_t d_len{0};
+  int d_id{0};
+
+  EVP_PKEY *d_edkey = nullptr;
+};
+
+bool OpenSSLEDDSADNSCryptoKeyEngine::checkKey() const
+{
+  return (d_edkey != nullptr);
+}
+
+void OpenSSLEDDSADNSCryptoKeyEngine::create(unsigned int bits)
+{
+  auto pctx = EVP_PKEY_CTX_new(d_edkey, nullptr);
+  if (pctx == nullptr) {
+    throw runtime_error(getName()+" context initialization failed");
+  }
+  if (EVP_PKEY_keygen_init(pctx) < 1) {
+    throw runtime_error(getName()+" keygen initialization failed");
+  }
+  if (EVP_PKEY_keygen(pctx, &d_edkey) < 1) {
+    throw runtime_error(getName()+" key generation failed");
+  }
+}
+
+DNSCryptoKeyEngine::storvector_t OpenSSLEDDSADNSCryptoKeyEngine::convertToISCVector() const
+{
+  storvector_t storvect;
+  string algorithm;
+
+#ifdef HAVE_LIBCRYPTO_ED25519
+  if(d_algorithm == 15) {
+    algorithm = "15 (ED25519)";
+  }
+#endif
+#ifdef HAVE_LIBCRYPTO_ED448
+  if(d_algorithm == 16) {
+    algorithm = "16 (ED448)";
+  }
+#endif
+  if (algorithm.empty()) {
+    algorithm = " ? (?)";
+  }
+
+  storvect.push_back(make_pair("Algorithm", algorithm));
+
+  string buf;
+  size_t len = d_len;
+  buf.resize(len);
+  if (EVP_PKEY_get_raw_private_key(d_edkey, reinterpret_cast<unsigned char*>(&buf.at(0)), &len) < 1) {
+    throw runtime_error(getName() + " Could not get private key from d_edkey");
+  }
+  storvect.push_back(make_pair("PrivateKey", buf));
+  return storvect;
+}
+
+std::string OpenSSLEDDSADNSCryptoKeyEngine::sign(const std::string& msg) const
+{
+  auto pctx = EVP_PKEY_CTX_new(d_edkey, nullptr);
```

Comment:
I don't think we need `pctx` at all here, `EVP_DigestSignInit()` will accept a `nullptr` instead.